### PR TITLE
ENH: sparse: add nonzero functionality to `min, max, argmin, argmax`

### PR DESCRIPTION
**Pull request by quantresearch1**
DISCLAIMER: This is simply the PR from [dloney:sparse_nonzero](https://github.com/dloney/scipy/tree/sparse_nonzero), gh-11899, merged with main. I have asked on his PR if I could pick it up given it's been stale for two years now. All credits goes to him.

Added nonzero parsing functionality into sparse matrix min, max, argmin, and argmax functions.

What does this implement/fix?
This allows a user to selectively consider only the nonzero values in the sparse min, max, argmin, and argmax functions. A programmer will often use a sparse matrix when zeros do not matter. Previous to this change, the programmer will have needed to mathematically manipulate the sparse matrix so that these functions gave the correct result when considering nonzero values. By passing in the nonzero keyword, these functions are now able to consider only nonzero values within the matrix. The change defaults to the existing functionality and only activates these changes if the user activates the input keyword.

This issue exists in several of my existing programs. As a benchmark, this implementation improves performance between 1.15x and 1.35x compared to manipulating the sparse matrix to identify nonzero values.

Additional information
This broadly mirrors the same functionality as implemented in cupy, which should help the compatibility of the packages.

### CODE REVIEW

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Quite a few lint errors (can check these locally with `tools/lint_diff.py`):
```
scipy/sparse/_data.py:358:1: W293 blank line contains whitespace
scipy/sparse/tests/test_base.py:3483:80: E501 line too long (81 > 79 characters)
scipy/sparse/tests/test_base.py:3512:80: E501 line too long (81 > 79 characters)
scipy/sparse/tests/test_base.py:3514:80: E501 line too long (80 > 79 characters)
scipy/sparse/tests/test_base.py:3521:80: E501 line too long (88 > 79 characters)
scipy/sparse/tests/test_base.py:3522:80: E501 line too long (88 > 79 characters)
scipy/sparse/tests/test_base.py:3525:80: E501 line too long (89 > 79 characters)
scipy/sparse/tests/test_base.py:3526:80: E501 line too long (89 > 79 characters)
scipy/sparse/tests/test_base.py:3535:80: E501 line too long (83 > 79 characters)
scipy/sparse/tests/test_base.py:3536:80: E501 line too long (83 > 79 characters)
scipy/sparse/tests/test_base.py:3544:80: E501 line too long (83 > 79 characters)
scipy/sparse/tests/test_base.py:3545:80: E501 line too long (83 > 79 characters)
```

**Comment by quantresearch1** [comment]
Thanks j-bowhay will fix now

**Comment by tupui** [comment]
Thank you @quantresearch1 could you link the previous PR this originates from?

**Comment by quantresearch1** [comment]
@j-bowhay SHould be fixed now. The remaining lint errors are not from this PR:
scipy/io/_idl.py:800:80: E501 line too long (91 > 79 characters)
scipy/sparse/tests/test_base.py:417:80: E501 line too long (82 > 79 characters)
scipy/sparse/tests/test_base.py:479:80: E501 line too long (82 > 79 characters)
scipy/sparse/tests/test_base.py:541:80: E501 line too long (82 > 79 characters)
scipy/sparse/tests/test_base.py:599:80: E501 line too long (82 > 79 characters)

Would you like to address these as well in this PR ? I am happy to do so but could be better separate.

**Comment by quantresearch1** [comment]
@tupui Happy to do so, I mentioned at the beginning this is originating from [dloney:sparse_nonzero](https://github.com/dloney/scipy/tree/sparse_nonzero) (which seem stale now). Is there a specific process to 'link' two PR in GItHub ? I have had a look online and the resources I find talk about linking a PR to an issue, not two PR together.

**Comment by tupui** [comment]
You just need to comment with the PR number with either `gh-` or `#` before the number.

**Comment by quantresearch1** [comment]
I think I understand now, the link I shared is not the link to the PR. Here is @dloney original PR:
https://github.com/scipy/scipy/pull/11899

**Comment by tupui** [comment]
> Would you like to address these as well in this PR ? I am happy to do so but could be better separate.

Please don't update code which you don't need to change for the PR. We want minimal diffs to retain meaningful blame information.

**Comment by quantresearch1** [comment]
> > Would you like to address these as well in this PR ? I am happy to do so but could be better separate.
> 
> Please don't update code which you don't need to change for the PR. We want minimal diffs to retain meaningful blame information.

Makes sense, I prefer it that way actually.

**Comment by dloney** [comment]
Thanks for picking this back up. Let me know if I can do anything to help get this finalized.

**Comment by tupui** [comment]
Mmm note that @dloney did not reply. I would wait a little bit. In the meantime, it seems like it was asked that an email should be send to the mailing list to solicit feedback. You're welcome to do so.

**Comment by quantresearch1** [comment]
> Thanks for picking this back up. Let me know if I can do anything to help get this finalized.
I am happy to take this PR down if you want to continue yours from two years ago

**Comment by dloney** [comment]
Happy to continue on this pr. It’s great you have bandwidth to push this forward since its lost momentum.

**Comment by tupui** [comment]
Thank you for clarifying things. I will close the other one then.

**Comment by quantresearch1** [comment]
Thanks @dloney , @tupui . Mail sent to the scipy dev team, will update with their feedback

**Comment by quantresearch1** [comment]
Your message to the SciPy-Dev mailing-list was rejected for the following
reasons:

The message is not from a list member

**Comment by quantresearch1** [comment]
I think I got there in the end: https://mail.python.org/archives/list/scipy-dev@python.org

**Comment by tupui** [comment]
All good now. First email from new contributors has to be validated.

File: scipy/sparse/tests/test_base.py [hunk 1]
```diff
@@ -3461,6 +3461,167 @@ def test_argmax(self):
             assert_raises(ValueError, mat.argmax, axis=axis)
             assert_raises(ValueError, mat.argmin, axis=axis)
 
+    def test_minmax_explicit_axis(self):
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py:3464]: I think these tests would be easier to understand if they were interleaved in the existing tests. That is, update the `test_argmax` method to add more `assert_equal()` calls for the `explicit=True` case instead of adding a new `test_argmax_explicit` method.

File: scipy/sparse/tests/test_base.py [hunk 2]
```diff
@@ -3461,6 +3461,167 @@ def test_argmax(self):
...
+                np.zeros((1, 0)),
+                X.min(axis=axis, explicit=True).A
+            )
+            assert_array_equal(
+                np.zeros((1, 0)),
+                X.max(axis=axis, explicit=True).A
+            )
+
+    def test_argmax_explicit(self):
+        D1 = np.array([[-1, 5, 2, 3],
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py:3586]: This test doesn't appear to be exercising the case where a row / column is all zero. We should make sure we cover both of those thoroughly.

File: scipy/sparse/_data.py [hunk 3]
```diff
@@ -273,6 +278,11 @@ def _arg_min_or_max(self, axis, out, op, compare):
                 am = op(mat.data)
                 m = mat.data[am]
 
+                if explicit:
+                    if m == 0:
+                        return []
```
Comment (perimosocordiae) [inline scipy/sparse/_data.py:283]: This seems wrong to me. We'll only get `m == 0` in the case where an explicit zero was stored, and it seems to me that an explicit zero is a valid index for an argmin/argmax operation to return, even with `explicit=True`.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

File: scipy/sparse/tests/test_base.py (hunk 1 above)
Comment (quantresearch1) [reply scipy/sparse/tests/test_base.py]: Done

File: scipy/sparse/tests/test_base.py (hunk 2 above)
Comment (quantresearch1) [reply scipy/sparse/tests/test_base.py]: Done

File: scipy/sparse/_data.py [hunk 4]
```diff
@@ -370,15 +389,19 @@ def argmax(self, axis=None, out=None):
...
         Returns
         -------
         ind : numpy.matrix or int
             Indices of maximum elements. If matrix, its size along `axis` is 1.
         """
-        return self._arg_min_or_max(axis, out, np.argmax, np.greater)
+        return self._arg_min_or_max(axis, out, np.argmax, np.greater, explicit)
 
-    def argmin(self, axis=None, out=None):
+    def argmin(self, axis=None, out=None, explicit=False):
```
Comment (tupui) [inline scipy/sparse/_data.py:404]: New arguments should be keywords only.

```suggestion
    def argmin(self, axis=None, out=None, *, explicit=False):
```

File: scipy/sparse/_data.py [hunk 5]
```diff
@@ -353,9 +372,9 @@ def min(self, axis=None, out=None):
         numpy.matrix.min : NumPy's implementation of 'min' for matrices
 
         """
-        return self._min_or_max(axis, out, np.minimum)
+        return self._min_or_max(axis, out, np.minimum, explicit)
 
-    def argmax(self, axis=None, out=None):
+    def argmax(self, axis=None, out=None, explicit=False):
```
Comment (tupui) [inline scipy/sparse/_data.py:377]:
```suggestion
    def argmax(self, axis=None, out=None, *, explicit=False):
```

File: scipy/sparse/_data.py [hunk 6]
```diff
@@ -321,9 +335,9 @@ def max(self, axis=None, out=None):
         numpy.matrix.max : NumPy's implementation of 'max' for matrices
 
         """
-        return self._min_or_max(axis, out, np.maximum)
+        return self._min_or_max(axis, out, np.maximum, explicit)
 
-    def min(self, axis=None, out=None):
+    def min(self, axis=None, out=None, explicit=False):
```
Comment (tupui) [inline scipy/sparse/_data.py:340]:
```suggestion
    def min(self, axis=None, out=None, *, explicit=False):
```

**Review by tupui — CHANGES_REQUESTED** [review]

File: scipy/sparse/_data.py (hunk 3 above)
Comment (quantresearch1) [reply scipy/sparse/_data.py]: Thanks @perimosocordiae for the review.  I have a question regarding the behaviour we want for argmin/max when explicit is True: at the moment, we have
        assert_equal(mat.argmin(explicit=True), np.argmin(mat.data))
        assert_equal(mat.argmax(explicit=True), np.argmax(mat.data))
If that's correct, we should raise a ValueError when axis is Noneand explicit is True for zero matrices because mat.data would be an empty array in that case. I have changed the code to do that but let me know if I misunderstood

Comment (quantresearch1) [reply scipy/sparse/_data.py]: I have added a simple benchmark for argmax just to make sure my changes does not have adverse impact:
Before changes:
[100.00%] ··· sparse.Argmax.time_argmax                                                   ok
[100.00%] ··· ========= ============= ============= =============
              --                          format                 
              --------- -----------------------------------------
               density       csr           csc           coo     
              ========= ============= ============= =============
                 0.01      699±50μs    1.10±0.02ms    14.0±0.2μs 
                 0.1     4.88±0.07ms   7.77±0.03ms     137±4μs   
                 0.5      24.0±0.5ms    34.2±0.9ms   2.37±0.05ms 
              ========= ============= ============= =============
After changes:
[100.00%] ··· sparse.Argmax.time_argmax                                                   ok
[100.00%] ··· ========= ============= ============= ============
              --                         format                 
              --------- ----------------------------------------
               density       csr           csc          coo     
              ========= ============= ============= ============
                 0.01      100±2μs      97.0±0.5μs   12.6±0.4μs 
                 0.1       439±3μs       437±3μs      64.9±3μs  
                 0.5     2.43±0.05ms   2.43±0.02ms    315±10μs  
              ========= ============= ============= ============
I think the time improvement are simply coming from the removal of sum_duplicates() which looks expensive

Comment (quantresearch1) [reply scipy/sparse/_data.py]: @perimosocordiae Regarding the test case: I mentioned above
assert_equal(mat.argmin(explicit=True), np.argmin(mat.data))
I now realise I forgot to take into account the non canonical case where it makes sense to return 'am'. I have added a simple test with a 2 by 2 zero matrix to make sure this works.

**Comment by yfarjoun** [comment]
just want to say that I'm excited that this PR might be happening....sparse-matrices need this!

**Comment by stefanv** [comment]
Two issues to decide on:

1. Is `explicit` the right naming for indicating that we're operating on assigned values?
2. Should `min` return 0 on rows without sparse values? My gut feel is no.

File: scipy/sparse/_data.py [hunk 7]
```diff
@@ -238,21 +247,25 @@ def _arg_min_or_max_axis(self, axis, op, compare):
             indices = mat.indices[p:q]
             am = op(data)
             m = data[am]
-            if compare(m, zero) or q - p == line_size:
-                ret[i] = indices[am]
+            if explicit:
+                if q - p > 0:
+                    ret[i] = indices[am]
```
Comment (jjerphan) [inline scipy/sparse/_data.py:252]: As indicated by @stefanv in https://github.com/scipy/scipy/pull/16467/#issuecomment-1563457909, we need to treat the case of an empty row or column (the missing `else` case here).

**Review by jjerphan — COMMENTED** [review]
A few comments. The main concern is the treatment of empty rows or columns and tests that are a bit hard to read.

Also, the introduction of a new key-word argument is relevant, yet I am not sure if `explicit` is the right name for it.

File: scipy/sparse/_data.py (hunk 7 above)
Comment (perimosocordiae) [reply scipy/sparse/_data.py]: One option is to return a sparse result.

Comment (jjerphan) [reply scipy/sparse/_data.py]: This looks like the appropriate solution to me. What do you think of it, @stefanv?

The COO format is probably appropriate here.

Comment (stefanv) [reply scipy/sparse/_data.py]: Yes, I like that solution!

**Comment by jjerphan** [comment]
@quantresearch1: are you interested to continue this PR? :slightly_smiling_face:

File: scipy/sparse/_data.py [hunk 8]
```diff
@@ -206,19 +215,19 @@ def _min_or_max(self, axis, out, min_or_max):
             if self.nnz == 0:
                 return zero
```
Comment (yfarjoun) [inline scipy/sparse/_data.py:216]: leaving this at return zero implies that the non-explicit elements have value zero...perhaps return None in this case?

**Comment by yfarjoun** [comment]
need to update the PR message "non-zero"->explicit

File: scipy/sparse/tests/test_base.py [hunk 9]
```diff
@@ -3467,19 +3552,76 @@ def test_argmax(self):
...
+
+        D = np.zeros((2, 2))
+        mat = self.spmatrix(D)
+        if mat.nnz != 0:
+            # Noncanonical case
+            assert_equal(mat.argmin(axis=None, explicit=True), 0)
+            assert_equal(mat.argmax(axis=None, explicit=True), 0)
+        else:
+            assert_raises(ValueError, mat.argmin, axis=None, explicit=True)
+            assert_raises(ValueError, mat.argmax, axis=None, explicit=True)
```
[the remark this answers is not in the document]
Comment (quantresearch1) [reply scipy/sparse/tests/test_base.py]: Just checking:  do you mean a breakdown of explicit rather than using even_explicit_pairs and odd_explicit_pairs ?
        for axis in axes_even:
            assert_raises(ValueError, mat.argmin, axis=axis, explicit=False)
            assert_raises(ValueError, mat.argmax, axis=axis, explicit=False)
            assert_raises(ValueError, mat.argmin, axis=axis, explicit=True)
            assert_raises(ValueError, mat.argmax, axis=axis, explicit=True)

File: scipy/sparse/_data.py [hunk 10]
```diff
@@ -154,13 +154,21 @@ def _find_missing_index(ind, n):
...
 
 
+def groupby(values):
+    diff = np.concatenate(([1], np.diff(values)))
+    idx = np.concatenate((np.where(diff)[0], [len(values)]))
+    index = np.empty(len(idx)-1, dtype='u4,u2')
+    index['f0'] = values[idx[:-1]]
+    index['f1'] = np.diff(idx)
+    return index
+
```
[the remark this answers is not in the document]
Comment (quantresearch1) [reply scipy/sparse/_data.py]: Can't remember what this was, I have deleted it

File: scipy/sparse/_data.py [hunk 11]
```diff
@@ -308,6 +327,12 @@ def max(self, axis=None, out=None):
             compatibility reasons. Do not pass in anything except
             for the default value, as this argument is not used.
 
+        explicit : {False, True} optional
+            Indicates whether the nonzero entries of the matrix should be
+            considered. If a row/column is empty, a zero will be returned
+            to indicate it contains no nonzero values.
```
[the remark this answers is not in the document]
Comment (quantresearch1) [reply scipy/sparse/_data.py]: Done. I have set 1.15 for the version placeholder but will correct that if the PR gets closed to being merged

**Comment by perimosocordiae** [comment]
I just resolved the merge conflict related to the new nanmin/nanmax methods. I added the `explicit` kwarg to those methods as well in the process.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]
This is getting very close to merge-ability (after much discussion, thanks to all involved!)

The one outstanding issue I'd like to see resolved is the matter of how to handle empty row/column results. I agree with @jjerphan that returning a COO-format sparse result would make sense, though until gh-18530 lands this will be a bit awkward, shape-wise.

Maybe the right solution is to change `explicit` from a boolean parameter to something like
 - `explicit=None`: current scipy behavior, including implicit zeros,
 - `explicit='dense'`: this PR's behavior with zeros returned for empty row/column data
 - `explicit='sparse'`: the proposed sparse return

**Comment by yfarjoun** [comment]
The explicit: "dense" option leads to very strange behaviour. If we agree that missing values could correspond to non-zero values, zero shouldn't be returned as a default value. if we want a dense option, the user should supply the value to use when there's no data. 

In my usecase, the sparse matricies are log-probabilities, so missing values are effectively -Inf. but if I expect the min/max to be dense, I'm happy to supply -Inf as the real missing value (though I'll probably just use 'sparse').

**Comment by yfarjoun** [comment]
that said, for argmin/argmax there's an interesting question as to whether an implicit value could ever be the min/max...if so, again, the user should provide the implicit value for comparing with the explicit values.

**Comment by dschult** [comment]
I am just coming to this discussion, and hope to nudge it over the line now that sparse arrays support 1D.

TL;DR
1) How about `only_stored` instead of `explicit` for the keyword name?
2) How about using the `format` keyword with value "dense", "sparse" or 3-letter format type?

**keyword name:**
- "explicit zeros", "stored values", "stored elements", "stored entries" are all terms used in SciPy to refer to values explicitly stored in the sparse array. The term "explicit" seems to be used only as an adjective for "zeros", while "stored" is used to describe zero and nonzero values. So I think "stored" or "explicit" are the two terms to choose from. 
- This new keyword is restricting the objects considered to only the explicit/stored. The word "explicit" by itself doesn't describe what is explicit nor what "True" would mean for this function. So I think "only" or some other indicator of limiting the min/max search would be appropriate.
- The choice between values/entries/elements is subjective, but 
    - "stored elements" is used in str/repr so appears many times in the repo (but only once outside of `str/repr`)
    - "stored values" is used five times in the repo -- more than the others, and often in regards to `getnnz` and what it counts.
    - "stored entries" only appears once in the repo, but "entries" is used often in "duplicate entries".
  I like "stored values", and we could change the word used in str/repr, but that changed recently for v1.14. So "elements" might be better.
- `only_stored_values` is a fairly complete description, but seems long.

For me, "stored" wins over "explicit" because "explicit" refers to zeros more than to general values. And I like a shorter version that avoids the elements/values/entries mix.

**Dense vs Sparse: `format=`**
How do we specify whether we want a dense or sparse result? This is a broader question than min/max, applying to at least `sum/mean` and arguably to any reduction function.  In discussing this with CJ, he mentioned that we could use the existing `format` keyword as a general way to indicate return type by adding a option `format="dense"`. The keyword can take on any of the 3-letter format values, "dense", and in some cases it might take on "sparse" to mean "any sparse format". The default value will likely vary by function, e.g. "dense" for `argmin` and "coo" for `min` (or maybe just "sparse" for min?).

Adding a `format` keyword to `min/max/argmin/argmax` would allow `only_stored` to be boolean, separate from the choice of dense or sparse return values. For backward compatibility, the default (None?) would indicate sparse COO for the min/max functions, and "dense" for the argmin/argmax functions.

**Comment by quantresearch1** [comment]
Hi Dan,Thanks for the feedback. I won’t have time to get this over the line in the end. Feel free to take it over.Kind regards,ArnaudOn Jun 10, 2024, at 7:04 AM, Dan Schult ***@***.***> wrote:﻿
I am just coming to this discussion, and hope to nudge it over the line now that sparse arrays support 1D.
TL;DR

How about only_stored instead of explicit for the keyword name?
How about using the format keyword with value "dense", "sparse" or 3-letter format type?

keyword name:

"explicit zeros", "stored values", "stored elements", "stored entries" are all terms used in SciPy to refer to values explicitly stored in the sparse array. The term "explicit" seems to be used only as an adjective for "zeros", while "stored" is used to describe zero and nonzero values. So I think "stored" or "explicit" are the two terms to choose from.
This new keyword is restricting the objects considered to only the explicit/stored. The word "explicit" by itself doesn't describe what is explicit nor what "True" would mean for this function. So I think "only" or some other indicator of limiting the min/max search would be appropriate.
The choice between values/entries/elements is subjective, but

"stored elements" is used in str/repr so appears many times in the repo (but only once outside of str/repr)
"stored values" is used five times in the repo -- more than the others, and often in regards to getnnz and what it counts.
"stored entries" only appears once in the repo, but "entries" is used often in "duplicate entries".
I like "stored values", and we could change the word used in str/repr, but that changed recently for v1.14. So "elements" might be better.


only_stored_values is a fairly complete description, but seems long.

For me, "stored" wins over "explicit" because "explicit" refers to zeros more than to general values. And I like a shorter version that avoids the elements/values/entries mix.
Dense vs Sparse: format=
How do we specify whether we want a dense or sparse result? This is a broader question than min/max, applying to at least sum/mean and arguably to any reduction function.  In discussing this with CJ, he mentioned that we could use the existing format keyword as a general way to indicate return type by adding a option format="dense". The keyword can take on any of the 3-letter format values, "dense", and in some cases it might take on "sparse" to mean "any sparse format". The default value will likely vary by function, e.g. "dense" for argmin and "coo" for min (or maybe just "sparse" for min?).
Adding a format keyword to min/max/argmin/argmax would allow only_stored to be boolean, separate from the choice of dense or sparse return values. For backward compatibility, the default (None?) would indicate sparse COO for the min/max functions, and "dense" for the argmin/argmax functions.

**Comment by dschult** [comment]
I believe this PR for min/max on stored elements only needs a couple of decisions still:
Recall that we decided to have min/max return a 1D sparse array with no stored value for any row/column that doesn't have any stored elements.

1) What should argmin/argmax return for empty rows/columns?  Currently we return dense results.
2) Should the keyword arg be named `only_stored`? (or something else?) Currently it is `explicit`.

**Comment by dschult** [comment]
Summary of discussion at weekly sparse array meeting:
- to minimize backward-compatibility issues and because `argmin/max` are often used to index arrays, let's continue to return dense. Changing to sparse, the user would likely just use `toarray` which would make `0`s show up in the index anyway. So we will return zeros for axes with no explicit elements.
- the `kwarg` will be named `explicit` (which while not perfect, indicates some of the complexity of  the whole issue).

While writing this, I thought of another possible return value for an empty axes: the invalid index `self.shape[axes]`. This could be easily tested for should the user want that and there is no ambiguity between valid 0 returns and empty axes 0 returns.

**Review by dschult — APPROVED** [review]
OK... this looks ready to go.
AFAICT decisions have been made and implemented, all green checks.
I approve this PR.

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
Merged! Thanks, all for sticking with this one over the last few years 😄
